### PR TITLE
inheritance.rst: Illustrate the case of abstract derived contract not providing base constructor arguments

### PR DESCRIPTION
**Pull request by aathan**
The prior text left questions open about what this might look like, and how it might be used.

### CODE REVIEW

File: docs/contracts/inheritance.rst [hunk 1]
```diff
@@ -484,7 +484,31 @@ inheritance list or in modifier-style in the derived constructor.
 Specifying arguments in both places is an error.
 
 If a derived contract does not specify the arguments to all of its base
-contracts' constructors, it will be abstract.
+contracts' constructors, it must be declared abstract. In that case, when
+a contract derives from it, its declaration or its constructor must
+provide the necessary parameters for all base classes that haven't had
```
Comment (cameel) [inline docs/contracts/inheritance.rst]: Or it could be abstract too.

File: docs/contracts/inheritance.rst [hunk 2]
```diff
@@ -484,7 +484,31 @@ inheritance list or in modifier-style in the derived constructor.
 Specifying arguments in both places is an error.
 
 If a derived contract does not specify the arguments to all of its base
-contracts' constructors, it will be abstract.
+contracts' constructors, it must be declared abstract. In that case, when
+a contract derives from it, its declaration or its constructor must
```
Comment (cameel) [inline docs/contracts/inheritance.rst]: A declaration would be the whole contract. The text above already refers to this as "inheritance list" so we should stick to that term here:
```suggestion
a contract derives from it, its inheritance list or its constructor must
```

Comment (cameel) [inline docs/contracts/inheritance.rst]: This is a bit hard to parse because "contract" and "it" are used multiple times referring to different things. How about this?
```suggestion
another contract derives from it, that contract's declaration or constructor must
```

File: docs/contracts/inheritance.rst [hunk 3]
```diff
@@ -484,7 +484,31 @@ inheritance list or in modifier-style in the derived constructor.
...
+
+    // DerivedAbstract does not specify Base parameters
+    abstract contract DerivedAbstract is Base {
+        constructor() {}
+    }
+
+    // Concrete must specify Base parameters
+    contract Concrete is DerivedAbstract {
+        constructor(uint y) Base(y * y) {}
+    }
```
Comment (cameel) [inline docs/contracts/inheritance.rst]: I'm not sure if this example provides enough value here for how much space it takes. It's nearly identical to the one above. Maybe we could just merge the two examples by making `Derived2` abstract and introducing `Derived3`?

Comment (aathan) [reply docs/contracts/inheritance.rst]: Agreed.

So, I just need a little hand holding on how to deal with PRs. I didn't immediately find a good guide, but if you know of one, please point me to it.  Basically, I accepted your very first suggestion and now your second suggestion is "stale" (marked [outdated]) and I can't accept (i.e. [commit suggestion]).

As far as dealing with the PR, I'm not clear on the cli process for that. I have a shallow clone of ethereium/solidity and I did a `gh pr checkout 12872` ... lets say I make a whole bunch of edits ... I can't push to the original PR because I don't have the rights in its repo. What do I do?

Thanks in advance (or for telling me to RTFM somewhere!)

Comment (cameel) [reply docs/contracts/inheritance.rst]: This is really because you're trying to use the github UI pretty intensively here. Normally I only see people doing that for quick edits and for bigger ones it's better to work locally with git. Github's UI clearly has some rough edges for more extensive editing. Unfortunately I do not know a good guide for this workflow and I myself don't use it but if you prefer to do it that way, it's fine. I'll try to suggest some solutions to problems you encounter if I know how.

> Basically, I accepted your very first suggestion and now your second suggestion is "stale" (marked [outdated]) and I can't accept (i.e. [commit suggestion]).

If you applied one suggestion and it modified lines that another suggestion would touch, you have a conflict. I don't think github provides any UI-level way to resolve that but you don't really have to use that apply button. You can change the text in the PR in whatever way that's feasible. E.g. if github lets you change it in an editor, you can just manually copy over my suggestion.

> I can't push to the original PR because I don't have the rights in its repo. What do I do?

That's because you can only push to your fork (i.e. https://github.com/aathan/solidity). That's where your branch is anyway. The code will be added to the main repo (i.e. https://github.com/ethereum/solidity) when the PR is merged. While it's still a PR, its in your repo and you have full control over it.

Comment (aathan) [reply docs/contracts/inheritance.rst]: Yes, I was just surprised the the github cli `gh pr checkout 12872` left me locally checked out in the solidity repo. I guess that's because I ran that command while in a shallow clone of the solidity repo. I'll try running that in a clone of my repo instead, but I think that didn't work. I'll follow up with you on the dev chat instead of polluting this conversation further.

**Review by cameel — APPROVED** [review]
This looks fine now so I'm approving.

We normally require review fixes to be squashed but since you're having trouble doing that via github UI I went ahead and squashed them for you. I also added some small tweaks directly to avoid going through another review round:
- Extended the example a bit to what I meant initially
- Added ``` `` ``` where needed.
- Added more index keywords.
- Removed trailing whitespace (`chk_coding_style` CI job was failing because of that)
- Slightly tweaked wording.

Please take a look and feel free to adjust if you feel these tweaks are not to your liking but if the text is acceptable, this PR could be merged as is (well. it'll still have to be rebased on #12868 unless @chriseth just force-merges it).

**Comment by aathan** [comment]
Sounds good. Please be explicit if any of what you said -- e.g. rebasing -- is my responsibility. I guess I would pull the patch out of my repo, do a git rebase, and push??

**Comment by cameel** [comment]
Yeah. We're using a rebase workflow which means that normally you'd create a branch on top of `develop` and keep it on top of it. I.e. when some PR gets merged into `develop`, you'd pull changes into your local working copy of the repo and do `git rebase` so that your branch is again on top.

You might get conflicts when rebasing, which means that the changes in `develop` touched the same parts you touched in your branch. In that case you need to resolve them. To do you select version from one of the branches or rewrite it to incorporate changes from both sides. Afterwards your branch looks as if you did your edits on top of the new version from `develop`. How you resolve them depends on the git tool your using.

We also usually want the commits in the branch to make it easy to understand what happened. I.e. if you changed something, then changed it back, then changed it again in a different way, we want just one commit going straight from the initial version to the final one. This is really more important with more complex code changes. For documentation changes this basically means that in most cases you should have just a single commit in your branch. If you have more, you can "squash" them into a single commit while rebasing. This also makes rebasing easier since the more commits, the more conflicts can there be. E.g. if you add something in one commit and remove it in another, you won't see it in the final diff - as if it was not touched at all - but individual commits can still conflict with other changes at that spot made in another PR merged in the meantime into `develop`.

**Comment by cameel** [comment]
Also, regarding this:

> Removed trailing whitespace (`chk_coding_style` CI job was failing because of that)

Random whitespace changes are just unnecessary noise in PR diffs and make reviewing harder. The typical way to avoid them is to have the editor automatically remove any trailing whitespace. Pretty much all programming editors have a setting for that.

One of the checks in the CI job that enforces coding style conventions is a check for trailing whitespace to make sure people have their editor configured like that. This check failed for you so I guess github's editor is not removing such whitespace?

**Comment by cameel** [comment]
> Added more index keywords.

As for index, I find it pretty handy when searching for stuff: https://docs.soliditylang.org/en/latest/genindex.html
Unfortunately not everyone remembers about updating these keywords. This time I noticed that after your changes some could be added so I went ahead and did it. This is not required but just nice to have. If you notice missing keywords in the section you're editing, please add them.

You can refer to [Index-generating markup](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#index-generating-markup) if you want to know the syntax but the gist that will get you through 99% of cases is that you put an index directive before a section and list topics, separated with commas:
```
.. index:: inheritance list, ! base;constructor, contract;abstract, abstract contract, ! snakes
```

keywords can contain spaces. `!` means that it's the "main" entry for this topic (otherwise it's just something on a different subject that happens to mention it). With a semicolon you create subentries for a broader topic - they'll show up as a small indented list under that keyword in the index.

**Comment by aathan** [comment]
Thx. I read up on rst syntax.

**Comment by cameel** [comment]
Looks like you closed the PR by accident.

It's because you renamed your branch `patch-16` to `clarify-try-catch-success-block`. Can you rename it back so that we can reopen it and merge? Unfortunately github does not let you rename branches that are already used in PRs. It just closes the PR if the original branch disappears.

**Comment by aathan** [comment]
Will do that now.

**Comment by cameel** [comment]
Hmm... looks like I still can't reopen this. If the "Reopen and comment" button is enabled for you, please use it. If not, maybe just create a new PR with the same branch.

**Comment by aathan** [comment]
New PR inbound.
